File: web/src/features/voice/sttSocket.ts
/**
 * One utterance over `WS /ws/stt`: PCM16LE, 16 kHz mono frames out, partial
 * transcripts back, one final transcript after `stop()`.
 *
 * A socket carries a single utterance because the gateway caps one at 120 s and
 * 4 MiB. Dictation longer than that is a chain of these, joined by
 * `TranscriptSegments`.
 */
import { socketUrl } from '../../lib/ws';

export type SttEvent =
  | { type: 'partial'; text: string }
  | { type: 'final'; text: string }
  | { type: 'failed'; message: string | null }
  | { type: 'closed' };

/** The slice of `WebSocket` this class uses, so a test can stand in for one. */
export interface SttSocketLike {
  binaryType: string;
  readyState: number;
  send(data: string | ArrayBuffer): void;
  close(): void;
  onopen: ((ev: unknown) => void) | null;
  onclose: ((ev: unknown) => void) | null;
  onerror: ((ev: unknown) => void) | null;
  onmessage: ((ev: { data: unknown }) => void) | null;
}

export type SttSocketFactory = (url: string) => SttSocketLike;

/** Gateway budget for one utterance: 4 MiB of audio. */
const MAX_AUDIO_BYTES = 4 * 1024 * 1024;

const OPEN = 1;

interface Options {
  language: string;
  onEvent: (event: SttEvent) => void;
  factory?: SttSocketFactory;
}

export class SttSocket {
  private socket: SttSocketLike | null = null;
  private sentBytes = 0;
  private finished = false;
  private opened = false;

  constructor(private readonly options: Options) {}

  /** Connect, and resolve once the gateway is ready for audio. */
  start(): Promise<void> {
    const factory = this.options.factory ?? ((url: string) => new WebSocket(url) as SttSocketLike);
    const url = `${socketUrl('/ws/stt')}?language=${encodeURIComponent(this.options.language)}`;
    const socket = factory(url);
    socket.binaryType = 'arraybuffer';
    this.socket = socket;
    socket.onmessage = (event) => this.receive(event.data);
    // Before the socket is open a failure rejects the connect; after it, the
    // same failure is an event for the dictation to weigh.
    return new Promise<void>((resolve, reject) => {
      socket.onopen = () => {
        this.opened = true;
        resolve();
      };
      socket.onerror = () => {
        if (this.opened) this.emit({ type: 'failed', message: null });
        else reject(new Error('stt socket failed'));
      };
      socket.onclose = () => {
        this.socket = null;
        if (this.opened) this.emit({ type: 'closed' });
        else reject(new Error('stt socket closed'));
      };
    });
  }

  /**
   * Append one capture buffer. Audio past the gateway's budget is dropped
   * rather than letting the socket be closed under the speaker mid-sentence.
   */
  append(frame: ArrayBuffer): void {
    const socket = this.socket;
    if (!socket || this.finished || socket.readyState !== OPEN) return;
    if (this.sentBytes + frame.byteLength > MAX_AUDIO_BYTES) {
      this.stop();
      return;
    }
    this.sentBytes += frame.byteLength;
    socket.send(frame);
  }

  /** Transcribe everything sent so far and wait for `stt.final`. */
  stop(): void {
    const socket = this.socket;
    if (!socket || this.finished) return;
    this.finished = true;
    if (socket.readyState === OPEN) socket.send(JSON.stringify({ type: 'stt.stop' }));
    else this.emit({ type: 'closed' });
  }

  /** Drop the utterance: no final transcript is wanted. */
  cancel(): void {
    const socket = this.socket;
    this.socket = null;
    this.finished = true;
    if (!socket) return;
    socket.onmessage = null;
    socket.onclose = null;
    socket.onerror = null;
    if (socket.readyState === OPEN) socket.send(JSON.stringify({ type: 'stt.cancel' }));
    socket.close();
  }

  private receive(data: unknown): void {
    if (typeof data !== 'string') return;
    let frame: { type?: string; text?: string; message?: string };
    try {
      frame = JSON.parse(data) as typeof frame;
    } catch {
      return;
    }
    if (frame.type === 'stt.partial') this.emit({ type: 'partial', text: frame.text ?? '' });
    if (frame.type === 'stt.final') this.emit({ type: 'final', text: frame.text ?? '' });
    if (frame.type === 'stt.error') this.emit({ type: 'failed', message: frame.message ?? null });
  }

  private emit(event: SttEvent): void {
    this.options.onEvent(event);
  }
}

File: ios/Sources/RCCore/Transport/STTSocket.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public enum STTEvent: Sendable, Equatable {
    case partial(String)
    case final(text: String, language: String)
    case failed(String)
    case closed
}

/// Streaming speech-to-text against the gateway.
///
/// The caller pushes PCM16LE, 16 kHz, mono frames; the gateway answers with a
/// partial transcript roughly every two seconds and one final transcript after
/// `stop()`. Audio leaves the phone here, which is the difference from
/// on-device recognition and is stated in the voice settings.
public actor STTSocket {
    public nonisolated let events: AsyncStream<STTEvent>

    /// Gateway limits: 120 s of audio, 4 MiB total per utterance.
    public static let maxAudioBytes = 4 * 1024 * 1024
    public static let maxAudioSeconds: TimeInterval = 120
    public static let sampleRate: Double = 16_000

    private let continuation: AsyncStream<STTEvent>.Continuation
    private let client: GatewayHTTPClient
    private let factory: any WebSocketFactory
    private let language: String
    private var connection: (any WebSocketConnection)?
    private var reader: Task<Void, Never>?
    private var sentBytes = 0
    private var finished = false
    private var receivedFinal = false
    private var deadline: Task<Void, Never>?

    /// How long the gateway gets to answer `stt.stop`. It has to transcribe up
    /// to 120 s of audio, so this is generous; it exists only so a gateway that
    /// never answers cannot leave the panel waiting forever.
    public static let finalTimeout: TimeInterval = 30

    public init(client: GatewayHTTPClient, language: String,
                factory: any WebSocketFactory = URLSessionWebSocketFactory()) {
        self.client = client
        self.language = language
        self.factory = factory
        let stream = AsyncStream<STTEvent>.makeStream(bufferingPolicy: .bufferingOldest(256))
        events = stream.stream
        continuation = stream.continuation
    }

    public func start() async throws {
        guard connection == nil, !finished else { return }
        guard let token = await client.bearerToken() else { throw TransportError.unauthorized }
        var components = URLComponents(url: client.endpoint.socketURL(path: "/ws/stt"), resolvingAgainstBaseURL: false)
        components?.queryItems = [URLQueryItem(name: "language", value: language)]
        guard let url = components?.url else { throw TransportError.invalidEndpoint }
        var request = URLRequest(url: url)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 30
        let socket = await factory.makeConnection(request: request)
        connection = socket
        await socket.resume()
        reader = Task { [weak self] in await self?.readLoop(socket) }
    }

    /// Append one capture buffer. Silently drops audio past the gateway budget
    /// rather than letting the socket be closed under the user mid-sentence.
    public func append(_ pcm: Data) async {
        guard let connection, !finished, !pcm.isEmpty else { return }
        guard sentBytes + pcm.count <= Self.maxAudioBytes else { await stop(); return }
        sentBytes += pcm.count
        do { try await connection.send(binary: pcm) } catch { await fail("audio upload failed") }
    }

    /// Finish and transcribe everything sent so far, then wait for `stt.final`.
    public func stop() async {
        guard let connection, !finished else { return }
        finished = true
        try? await connection.send(text: #"{"type":"stt.stop"}"#)
        deadline?.cancel()
        deadline = Task { [weak self] in
            try? await Task.sleep(for: .seconds(Self.finalTimeout))
            guard !Task.isCancelled else { return }
            await self?.giveUpWaiting()
        }
    }

    private func giveUpWaiting() async {
        guard !receivedFinal else { return }
        await fail(L10n.string("The gateway did not return a transcript."))
    }

    /// Discard the utterance. The gateway drops the audio and closes.
    public func cancel() async {
        let socket = connection
        finished = true
        try? await socket?.send(text: #"{"type":"stt.cancel"}"#)
        await close()
    }

    public func close() async {
        deadline?.cancel(); deadline = nil
        reader?.cancel(); reader = nil
        let socket = connection
        connection = nil
        await socket?.cancel()
        continuation.yield(.closed)
        continuation.finish()
    }

    private func readLoop(_ socket: any WebSocketConnection) async {
        while !Task.isCancelled {
            do {
                let data = try await socket.receive()
                guard let json = try? JSONDecoder().decode(JSONValue.self, from: data),
                      let object = json.objectValue, let type = object.string("type") else { continue }
                switch type {
                case "stt.partial":
                    continuation.yield(.partial(object.string("text") ?? ""))
                case "stt.final":
                    receivedFinal = true
                    continuation.yield(.final(text: object.string("text") ?? "",
                                              language: object.string("language") ?? language))
                    await close()
                    return
                case "stt.error":
                    await fail(object.string("message") ?? "Transcription failed.")
                    return
                default:
                    continue
                }
            } catch {
                guard !Task.isCancelled else { return }
                // A close without a final is the end of the utterance: whatever
                // the last partial produced is what the user gets.
                await fail(finished
                    ? L10n.string("The transcript did not finish. What was recognised is in your draft.")
                    : L10n.string("The transcription connection dropped."))
                return
            }
        }
    }

    private func fail(_ message: String) async {
        continuation.yield(.failed(message))
        await close()
    }
}
